### PR TITLE
Added Shogun's native dot impl to linalg under NATIVE backend

### PR DESCRIPTION
**Pull request by lambday**
Addresses the basics for issue #2667 and #2582

### CODE REVIEW

**Comment by iglesias** [comment]
It is looking good to me.

File: src/shogun/mathematics/linalg/linalg.h [hunk 1]
```diff
@@ -35,40 +35,40 @@
...
+ * not set explicitly via cmake options. A Native backend is also defined to
+ * accommodate Shogun's native implementation of some operations.
  *
  * The enum defines these backends in order of priority as default backend, as
- * in, first defined one will be used as default
- *
- * Note - Currently EIGEN3 is the default (if its available)
+ * in, first defined one will be used as default.
  *
+ * Note - Currently EIGEN3 is the default (if it is available).
```
Comment (iglesias) [inline src/shogun/mathematics/linalg/linalg.h]: Should this comment be updated?

Comment (lambday) [inline src/shogun/mathematics/linalg/linalg.h]: @iglesias umm I think no.. it is still true :D

**Comment by lambday** [comment]
@iglesias @lisitsyn I think we better merge this one _after_ we merge khaled's convolution patch #2509. I think he is not available for rebase. Since this patch also contains changes in Core.h I think we can get merge conflicts. I will then rebase and push again.

File: src/shogun/mathematics/linalg/linalg.h (hunk 1 above)
Comment (iglesias) [inline src/shogun/mathematics/linalg/linalg.h]: I'm afraid I am not completely understanding what initialisation of the last entry of the enum to 0 does. Aren't eigen and viennacl (if available) as well as native as default all getting the value zero?

Comment (lambday) [inline src/shogun/mathematics/linalg/linalg.h]: @iglesias if both are available, then the enum would be defined as
```c++
enum class Backend
{
    EIGEN3, // will be 0
    VIENNACL, // will be 1
    NATIVE, // will be 2
    DEFAULT = 0 // therefore EIGEN3
};
```
If eigen3 and viennacl is not found, the enum is
```c++
enum class Backend
{
    NATIVE, // will be 0
    DEFAULT = 0 // therefore NATIVE
};
```
We're relying on the fact that enum by default starts from 0 and keeps on increasing by 1.

Comment (lambday) [inline src/shogun/mathematics/linalg/linalg.h]: `DEFAULT` will be same as the first thing defined. So if both Eigen3 and ViennaCL are there, it is Eigen3. If only ViennaCL is there it is ViennaCL. If nothing is there it is NATIVE (which is limited to only a few operations that we'll provide native implementation).

Comment (sanuj) [inline src/shogun/mathematics/linalg/linalg.h]: Of what I understand,
- if we have eigen and viennacl then 
`EIGEN3=0, VIENNACL=1, NATIVE=2 and DEFAULT=0`
in which case Backend::DEFAULT is EIGEN3
- if we have eigen and not viennacl then
`EIGEN3=0, NATIVE=1, DEFAULT=0`
in which case `Backend::DEFAULT` is `EIGEN3`
- if we have viennacl and not eigen then
`VIENNACL=0, NATIVE=1, DEFAULT=0`
in which case `Backend::DEFAULT` is `VIENNACL`
- if we neither have eigen nor viennacl then,
`NATIVE=0, DEFAULT=0`
in which case `Backend::DEFAULT` is `NATIVE`

@lambday ?

Comment (lambday) [inline src/shogun/mathematics/linalg/linalg.h]: @sanuj yep. that's pretty much it :)

Comment (iglesias) [inline src/shogun/mathematics/linalg/linalg.h]: I understood the enum above now, @lambday. Thanks for the explanation!

Comment (lambday) [inline src/shogun/mathematics/linalg/linalg.h]: @iglesias thanks for merging the convolution patch. I assume it is okay then if I rebase and merge this patch then.

Please check the wiki on linalg as well. I could really use your suggestions on that.

Comment (iglesias) [inline src/shogun/mathematics/linalg/linalg.h]: Hey, @lambday. A couple of comments regarding the linalg wiki:

* When the Redux and Core modules are introduced [here](https://github.com/shogun-toolbox/shogun/wiki/README_linalg#relying-on-globally-set-linalg-backend), maybe it would be nice to mention what is different between both? Just to give an idea where the names come from.
* I saw the usage of linalg in CustomKernel.cpp. I wonder why the `HAVE_LINALG_LIB` guard is required. I guess it is because there is a cmake option that allows to switch on/off compilation of linalg. In that case, maybe that can be mention in the wiki as well.

Comment (iglesias) [inline src/shogun/mathematics/linalg/linalg.h]: It is very cool to have this documented in the wiki by the way. Awesome :-)

Comment (lambday) [inline src/shogun/mathematics/linalg/linalg.h]: @iglesias thanks for your comments.
- yes you're right, I'll add some comments on the module names. It is not very clear from those names. IMO redux is for those operations which takes a matrix/vector and returns a scalar, e.g. dot, sum etc. rest generic util things should go to core.
- actually the guard is not required. `HAVE_LINALG_LIB` is defined whenever we have `HAVE_EIGEN3` _or_ `HAVE_VIENNACL` defined. currently there is no cmake option to turn linalg off. I think only `HAVE_EIGEN3` check will suffice there. I'll take care of that.

Thanks. I hope the wiki will be useful :) will finish it off by tonight.

Comment (iglesias) [inline src/shogun/mathematics/linalg/linalg.h]: It sounds like we can get rid of `HAVE_LINALG_LIB` completely, right? With completely I mean everywhere and not only in CustomKernel.cpp. Operations in native will always be available so no guard required wherever they are used, and eigen3/viennacl stuff should always have their own have_eigen3/viennacl.

Comment (lambday) [inline src/shogun/mathematics/linalg/linalg.h]: @iglesias actually we don't want to implement _everything_ natively, e.g. we most certainly don't want to write our own self adjoint eigensolver or SVD/QR/Cholesky. `HAVE_LINALG_LIB` guard is useful for those methods which are available _only_ when we _do_ have an external linalg lib. In linalg modules we're guarding those operations with `HAVE_LINALG_LIB` which do not have their native implementations (this modification is done in this patch only). I think it is still useful. A use-case is when we just want the result for some of these operations and don't wanna bother about which backend was used (i.e. relying on global backend setup).

In CustomKernel.cpp I'm explicitly using eigen3 backend because when I added this, ViennaCL specializations were not added till then for those methods. We can most certainly do native implementation for `sum` methods, and, in that case, then we can get rid of the guard in CustomKernel. Otherwise, as of now, we can get rid of `HAVE_EIGEN3` guard and remove the specific backend Eigen3 in those `sum` methods - it will then use Eigen3/ViennaCL - whichever is available.

Comment (lambday) [inline src/shogun/mathematics/linalg/linalg.h]: I think I should mention about proper usage of `HAVE_LINALG_LIB` in the wiki as well as the limitations while using `NATIVE` backend more specifically. Thanks for your inputs :)

Comment (lambday) [inline src/shogun/mathematics/linalg/linalg.h]: I thought of putting a matrix there in the wiki - mentioning the currently supported operations and which backends we can use for that. I hope that would be helpful.

Comment (iglesias) [inline src/shogun/mathematics/linalg/linalg.h]: Oh, I got now the meaning of `HAVE_LINALG_LIB`; it refers to having an external library e.g. Eigen3 or ViennaCL. So far, I had read it as having the Shogun internal linalg library and that's why I got confused :-)

Comment (lambday) [inline src/shogun/mathematics/linalg/linalg.h]: Yeah the meaning is not very clear. I should explain this macro properly in the wiki :)
BTW I just rebased this branch. Shall we merge it after travis is finished?

Comment (lambday) [inline src/shogun/mathematics/linalg/linalg.h]: agh there are some errors due to c++11 code being exposed.. will fix it.

Comment (sanuj) [inline src/shogun/mathematics/linalg/linalg.h]: @lambday How much work is left here?

Comment (lambday) [inline src/shogun/mathematics/linalg/linalg.h]: @sanuj just addition of 4 lines. Sorry I just have been lazy P Will push by tonight.

Comment (sanuj) [inline src/shogun/mathematics/linalg/linalg.h]: Cool :) i have some holidays now so i would get time to work on the cleanup
and native implementations. That's why i was asking. Hope you guys merge
this soon.

Comment (lambday) [inline src/shogun/mathematics/linalg/linalg.h]: @sanuj yeah I guess @iglesias is fine with it (once travis stops whining). I'll merge this after I push the changes. Glad to see that you're psyched to work on this :D 

Go `linalg` !

**Comment by sanuj** [comment]
@lambday what's up? :)

**Comment by lambday** [comment]
Done. Travis failure not related. I'll ask @sonney2k if its okay to merge this patch right now.

File: src/shogun/mathematics/linalg/internal/modules/Core.h [hunk 2]
```diff
@@ -45,18 +47,18 @@ namespace shogun
...
  * @param C Result of the operation
  * @param transpose_A Whether to the transpose of A should be used instead of A
  * @param transpose_B Whether to the transpose of B should be used instead of B
- * @param overwrite If true, the values in C are overwritten with the result, 
+ * @param overwrite If true, the values in C are overwritten with the result,
  * otherwise, the result is added to the existing values
  */
 template <Backend backend=linalg_traits<Core>::backend,class Matrix>
-void matrix_product(Matrix A, Matrix B, Matrix C, 
+void matrix_product(Matrix A, Matrix B, Matrix C,
```
Comment (karlnapf) [inline src/shogun/mathematics/linalg/internal/modules/Core.h]: @yorkerlin consider to start using such things, as they become available

File: src/shogun/mathematics/linalg/internal/modules/SpecialPurpose.h [hunk 3]
```diff
@@ -67,15 +69,15 @@ void rectified_linear(Matrix A, Matrix result)
...
 	implementation::special_purpose::multiply_by_rectified_linear_derivative<backend, Matrix>::compute(A, C);
 }
 
-/** Applies the softmax function inplace to a matrix. The softmax function is 
- * defined as \f$ f(A[i,j]) = \frac{exp(A[i,j])}{\sum_i exp(A[i,j])} \f$ 
+/** Applies the softmax function inplace to a matrix. The softmax function is
+ * defined as \f$ f(A[i,j]) = \frac{exp(A[i,j])}{\sum_i exp(A[i,j])} \f$
  */
 template <Backend backend=linalg_traits<Core>::backend,class Matrix>
 void softmax(Matrix A)
```
Comment (karlnapf) [inline src/shogun/mathematics/linalg/internal/modules/SpecialPurpose.h]: @yorkerlin might be useful for the multiclass logistic GPs

**Comment by karlnapf** [comment]
GREAT work guys :)
@lambday will read your tutorial for linalg these days

File: src/shogun/mathematics/linalg/internal/modules/SpecialPurpose.h (hunk 3 above)
Comment (yorkerlin) [inline src/shogun/mathematics/linalg/internal/modules/SpecialPurpose.h]: Yes, I may use in near future once the four FITC methods are implemented.
There are two additional inference methods (all about Emti's KL dual inference method) to be added in the multiclass logistic GPs.  I will do it once I get time.

Comment (yorkerlin) [inline src/shogun/mathematics/linalg/internal/modules/SpecialPurpose.h]: BTW, @karlnapf  in theory, it seems FITC could work for the multiclass logistic GP. Not sure whether it is robust or not. Maybe, I take a look at this later.

**Comment by lambday** [comment]
@karlnapf thanks. Lot more to do. Is this a good time for merging a new patch? I'll merge it then.

**Comment by lambday** [comment]
Merging this one.

**Comment by lambday** [comment]
@sanuj you can now start working on the things you have planned. Please add one method per PR (easy to review). Proper unit testing must.

**Comment by sanuj** [comment]
Great! Yeah I'll keep that in mind @lambday Thanks!

File: src/shogun/mathematics/linalg/internal/modules/SpecialPurpose.h (hunk 3 above)
Comment (karlnapf) [inline src/shogun/mathematics/linalg/internal/modules/SpecialPurpose.h]: I guess my point was more to use the logit and softmax implementation from linalg in the CLogitLikelihood and CSoftMaxLikelihood. So can change existing implementations to use linalg

FITC for multiclass might not work so well, we have to check, but lets discuss somewhere else as off topic here.
